Create volunteer form

elineelorm opened the pull request:
I created a form to allow people register to be a volunteer.
I am still figuring out how to connect this page to the list of patients page.

sbahaud commented on app/src/main/java/com/google/sps/Volunteer.java:
```diff
@@ -0,0 +1,25 @@
+
+public final class Volunteer {
+
+    private final String firstName;
```
It is good to have a unique id associated with each volunteer. You can derive that from the email provided.

elineelorm replied:
I wasnt quite sure how to resolve this but I did add an id attribute to the volunteer in java

sbahaud commented on app/src/main/java/com/google/sps/Volunteer.java:
```diff
@@ -0,0 +1,25 @@
+    private final String lastName;
+    private final String email;
+    private final String phone;
+    private final String birthday;
```
Is birthday required. less personal info you collect from the volunteer, the better.

sbahaud commented on app/src/main/webapp/style.css:
```diff
@@ -280,4 +280,71 @@ patient {
+    display: none;
+}
+
+button {
```
Instead of changing the style for all buttons, maybe add a class to the button and add that style to the button element.

sbahaud commented on app/src/main/webapp/vol_Reg.html:
```diff
@@ -0,0 +1,59 @@
+
+        <div class="tab">Qualifications:
+            <p><input placeholder="Degree..." oninput="this.className = ''"></p>
+            <p><input placeholder="Volunteer Experience..." oninput="this.className = ''"></p>
```
It is not clear what info needs to be added here. Is it just free form text?

sbahaud commented on app/src/main/webapp/vol_Reg.html:
```diff
@@ -0,0 +1,59 @@
+        <div class="tab">Qualifications:
+            <p><input placeholder="Degree..." oninput="this.className = ''"></p>
+            <p><input placeholder="Volunteer Experience..." oninput="this.className = ''"></p>
+            <p><input placeholder="Reference Email..." oninput="this.className = ''"></p>
```
Not clear what Reference email is and how it is different from the "email" in the previous step.

sbahaud commented on app/src/main/webapp/vol_Reg.html:
```diff
@@ -0,0 +1,59 @@
+            <p><input type="checkbox" id="pw" onclick="showPassword()"> 
+                <label for="pw"> Show Password </label><br>
+            </p>
+        </div>
```
I would remove this for now. It would be better to use google signin for the signin, so you don't need to store username and passwd. You should be very careful with handling username/passwd.

sbahaud commented on app/src/main/webapp/vol_Reg.js:
```diff
@@ -0,0 +1,67 @@
+console.log(currentTab);
+
+function showTab(n) {
+var x = document.getElementsByClassName("tab");
```
Please indent the function properly.

sbahaud commented on app/src/main/webapp/vol_Reg.js:
```diff
@@ -0,0 +1,67 @@
+
+function validateForm() {
+var x, y, i, valid = true;
+x = document.getElementsByClassName("tab");
```
x is not a usefule variable. Please use variable names that clearly show that the variable is. Maybe use tabNum instead of x and tabInputs instead of y?

sbahaud commented on app/src/main/webapp/vol_Reg.js:
```diff
@@ -0,0 +1,67 @@
+}
+
+function fixStepIndicator(n) {
+var i, x = document.getElementsByClassName("step");
```
Here also rename x to steps maybe.

sbahaud reviewed (commented):
Added  few comments

elineelorm commented:
- Created the datastore
- Allowed saving info into datastore
- Linked the form to the patients.html in submit (It takes a while to load though)
- Added the various servlets
- Tried to use JSoup to filter input but didn't work
- Updated the variables with proper names

sbahaud commented on app/src/main/java/com/google/sps/servlets/FormHandlerServlet.java:
```diff
@@ -0,0 +1,51 @@
+    String email = request.getParameter("email");
+    String experience = request.getParameter("experience");
+    String reference = request.getParameter("reference"); 
+
```
Like I mentioned in our meeting, remember that any user input has to be validated and escaped to make sure there is no script. You can use jsoup for that. You can do that in a follow up PR

elineelorm replied:
I was unsure about how to do this so I'll try and do it in a new PR as suggested

sbahaud commented on app/src/main/java/com/google/sps/servlets/FormHandlerServlet.java:
```diff
@@ -0,0 +1,51 @@
+            .set("email", email)
+            .set("experience", experience)
+            .set("reference", reference)
+            .build();
```
Also the id. You can make it a random number.

elineelorm replied:
I made it the combination of a name and number

sbahaud commented on app/src/main/java/com/google/sps/Volunteer.java:
```diff
@@ -0,0 +1,34 @@
+        this.id = firstName + lastName;
+    }
+
+    public Volunteer(){
```
nit: add space before {

sbahaud commented on app/src/main/java/com/google/sps/Volunteer.java:
```diff
@@ -0,0 +1,34 @@
+        this.experience = "n/a";
+        this.reference = "n/a";
+        this.id = "n/a";
+    }
```
I would recommend, setting everything to empty instead of storing n/a. The client can then decide what to do when the string is empty.

sbahaud commented on app/src/main/java/com/google/sps/servlets/NewVolunteerServlet.java:
```diff
@@ -0,0 +1,56 @@
+
+    List<Volunteer> volunteers = new ArrayList<>();
+    if (results == null){
+        Volunteer volunteer = new Volunteer("n/a", "n/a", "n/a", "n/a", "n/a");
```
You can make it all empty and just do new Volunteer();

sbahaud reviewed (approved):
Approving with some comments. You can address the comments in a later PR
